RFC: contrast coding

kleinschmidt opened the pull request:
This PR introduces a system for contrast coding, or controlling how categorical variables are converted into columns in a `ModelMatrix`.  The basic interface is that the `ModelFrame` constructor takes an optional keyword argument `contrasts`, which is a `Dict` mapping column name symbols to contrasts, specified as subtypes of `AbstractContrast` or instances thereof.  This addresses (at least partially) #757 and #788.

The code could be cleaned up quite a bit and refactored, but I wanted to get feedback on the general approach so far.  It also still doesn't address the problem of needing to switch contrast coding based on whether or not lower-order effects are also present (also raised in #757), but I'm planning on also implementing those checks.

johnmyleswhite commented:
This is really timely as the issue of contrast coding came this weekend when talking with the folks at Dato about their DataFrame -> Matrix conversion tools.

Excited to review this.

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -0,0 +1,143 @@
+
+type TreatmentContrast <: AbstractContrast
+    base::Integer
+    matrix::Array{Any,2}
```
Probably shouldn't be `Any`, but rather a type parameter inferred from the arguments to the constructor.

kleinschmidt replied:
For the matrix, I think the type should actually be `Float64`, at least by analogy with how continuous columns are coerced to `Vector{Float64}`. But the types for `termnames` and `levels` probably should be inferred based on the underlying data (or that termnames is always going to be coerced to a string...)

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -0,0 +1,143 @@
+    lvls = levels(v)
+
+    n = length(lvls)
+    if n < 2 error("not enought degrees of freedom to define contrasts") end
```
"enough". In Julia, this kind of check is usually written `n < 2 || ...`.

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -0,0 +1,143 @@
+    ## are the same by constructing a re-indexing vector. Indexing into
+    ## reindex with v.refs will give the corresponding row number of the
+    ## contrast matrix
+    reindex = [findfirst(l .== contrast.levels) for l in levels(v)]
```
`findfirst(contrast.levels, l)` should work and be slightly more efficient.

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -0,0 +1,143 @@
+termnames(term::Symbol, col::Any, contrast::AbstractContrast) =
+    ["$term - $name" for name in contrast.termnames]
+
+function cols(v::PooledDataVector, contrast::AbstractContrast)
```
I'm not a fan of the name `cols`. I know it was already there, but could we find something more explicit?

kleinschmidt replied:
How about something like `modelmatrix_columns`?

nalimilan replied:
Yeah, why not.

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -0,0 +1,143 @@
+## Constructing a contrast from a non-pooled data vector will first pool it
+## 
+## (NOT SURE THIS IS GOOD: constructing columns also depends on levels, so for
+## _that_ to work would need to somehow hold onto pooled data...)
```
I don't understand this comment. Could you elaborate?

kleinschmidt replied:
Honestly, I'm not sure what I was thinking now when I wrote that :)  

There was some discussion in #867 about a mechanism for treating continuous variables as categorical. One way to do this would be to provide a contrast for that column, and a contrast constructor that takes a column that's _not_ a `PooledDataArray`, calls `pool` on the column before evaluating the contrast (like here). I think I was worried that somehow the levels of the new pooled vector wouldn't be held onto, but that's wrong (they're stored in the contrast object that's created when the `ModelFrame` constructor calls the contrast constructors).

The thing that's missing here is a `cols` method that takes a non-`PooledDataArray` column and a contrast and handles that. The problem there is that you're calling `pool` twice (at least). The alternative is somehow to store the pooled version of the column but that just seems really messy to me.

nalimilan replied:
Maybe we could avoid calling `pool` to create the contrasts? Just use `unique`, and allow passing an optional order of levels to `contrast!`. That could also be useful to specify `base` for cases where you want to use a different reference just for one model, without changing it for the PDA itself. IIUC in the current code of this PR, there's no way to pass `base` when calling `ModelFrame`.

You could imagine that `*Contrast` types, instead of actually requiring the data to be passed to their constructors, would just hold the levels in the specified order or the index of the base level, or nothing to keep defaults. Then, under the hood another function would combine the data and the contrast info, and create the contrast matrix.

So you could write things like:
```julia
ModelFrame(Formula(Nothing(), :x), d, contrasts = [:x => SumContrast])
# Equivalent
ModelFrame(Formula(Nothing(), :x), d, contrasts = [:x => SumContrast()])

# With non-default values
ModelFrame(Formula(Nothing(), :x), d, contrasts = [:x => SumContrast(base=2)])
ModelFrame(Formula(Nothing(), :x), d, contrasts = [:x => SumContrast(levels=["A", "C", "B"])])
```
```

kleinschmidt replied:
You can specify a base just for one model frame by doing something like

```julia
ModelFrame(Formula(Nothing(X), :x), d, contrasts = [:x => SumContrast(d[:x]; base=2)])
```

A little clunky, to be sure.  Another possibility would be to allow calling the constructor without data (but with the base level or other non-default arguments), and overloading call to accept a data vector and creating/storing the contrast matrix then. That would cause trouble though if we make the type of the levels a type parameter of the contrast types, since it wouldn't be possible to know the determine the parameter without the data...

kleinschmidt replied:
Also, what you suggested is basically the first implementation I thought of, too. But it has portability problems: without guaranteeing that an instantiated contrast has the contrast matrix, you can't guarantee that the contrast will always produce the same number of columns. 

Also also, for what it's worth, specifying the levels and the base level is really all that's required to fully specify a contrast. I'd be perfectly happy with also having constructors that directly take the levels and the base level. And in fact the constructor that takes a `PooledDataArray` can just call the levels-based constructor...

nalimilan replied:
How important is it to be sure an instance of `*Contrasts` doesn't change the number of columns? On the contrary, it seems like a feature that you could do e.g. `
ModelFrame(Formula(Nothing(), :x), d, contrasts = [:x => SumContrast(base=2)])` and still apply the same model frame on another data set, possibly with different levels. This wouldn't be possible if you do `ModelFrame(Formula(Nothing(), :x), d, contrasts = [:x => SumContrast(d[:x]; base=2)])`.

There could be another type, for example `*ContrastMatrix`, which would store the actual matrix, and would only be computed when the levels are known. That would logically parallel the `ModelFrame`/`ModelMatrix` pair. The type stability issue would also go away: only `*ContrastMatrix` would be parameterized on the type of the data.

kleinschmidt replied:
I don't actually think that it's currently possible to apply an existing `ModelFrame` to different data.  A `ModelFrame` is essentially a `Formula` (or rather it's parsed form as `Terms`) plus a `DataFrame`. It's the `Formula`/`Terms` that are portable over data. Perhaps that's actually an argument for storing contrasts in the `Formula`/`Terms` (rather than the `ModelFrame`)...

Also, if we allow users to specify the base _level_ (rather than the index as currently) I'm not sure it even makes sense to allow that in the absence of data (since you'd want to validate it against the actual levels present in the data...)

nalimilan replied:
Right. So it would make sense to specify the contrasts directly in the formula rather than in the call to `ModelFrame`. In R, the `C` function sets the contrasts as an attribute on factors (equivalent to PDAs), and as a standard function it can be used either separately, or directly in the formula. This solution implies storing an additional field for each PDA. Though it should be negligible if we only store a reference to the contrast type (and possibly the reference level), so it might be suited for the `CategoricalVariable` replacement for PDAs. An alternative would be to create a wrapper around a vector, which would hold both the data and the contrasts.

An advantage of these two strategies is that you would be able to specify the contrasts for a given variable once for all, which is a quite common need AFAICT. And the point about specifying the base level before the data is available would go away.

kleinschmidt replied:
I'd thought about creating something like a `ContrastedPooledDataArray` like you describe, that wraps a PDA, but then looked at the `PooledDataArray` API and it seemed like more trouble than it was worth to create a wrapper type and delegate everything. Also, it seems like it might be prudent to see how categorical data implementation shakes out before marrying contrasts and categorical data itself, although I'm not sure what the status of that is at the moment (perhaps @johnmyleswhite wants to weigh in here...)

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -0,0 +1,143 @@
+    if n < 2 error("not enought degrees of freedom to define contrasts") end
+
+    not_base = [1:(base-1); (base+1):n]
+    tnames = lvls[ not_base ]
```
Spacing is unusual here.

nalimilan commented on src/DataFrames.jl:
```diff
@@ -36,6 +37,9 @@ export @~,
        ModelFrame,
        ModelMatrix,
        SubDataFrame,
+       SumContrast,
```
Shouldn't the name be plural?

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -0,0 +1,143 @@
+abstract AbstractContrast
+
+termnames(term::Symbol, col::Any, contrast::AbstractContrast) =
+    ["$term - $name" for name in contrast.termnames]
```
`"$term $name"` or `"$term.$name"` (or something else) looks like a more common convention to me. `-` could mean "minus" in the context of a regression model.

kleinschmidt replied:
I think the R convention is something like "$term$name" which I really don't like. Where have you seen those? How about something like "$term: $name"?

nalimilan replied:
Right, adding a separator isn't a bad idea. `"$term: $name"` is good, though we need to think about how interactions will appear. For example: `x: A&y` would naturally be read as  `x: (A&y)`, instead of `(x: A)&y`; it would be even worse with interactions between categorical variables. Removing the space after the semicolon might be enough. (The bikeshedding session is open!)

kleinschmidt replied:
At the moment, the interaction separator has spaces (so you'd get `x: A & y`).  Another possibility is `x (A)` and `x (A) & y`.

nalimilan commented:
Thanks, this is quite nice! I've only made superficial comments, but the global strategy sounds good to me -- though I'm not a specialist of this code at all.

A feature I've been dreaming of for some time, and which R doesn't have, would be to have the ability to compute the redundant/omitted coefficient after the model has been estimated. This would be extremely useful to print tables for publications. For example, for treatment contrasts, to get the name of the reference level, and the associated coefficient (0). For sum contrasts, it would consist in getting the name of the omitted level, with the coefficient equal to the sum of all coefficients estimated in the model for other levels. A method for each contrasts type could provide this information.

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -0,0 +1,126 @@
+## Could write this as a macro, too, so that people can register their own
+## contrast types easily, without having to write out this boilerplate...the
+## downside of that would be that they'd be locked in to a particular set of
+## fields...although they could always just write the boilerplate themselves...
```
The boilerplate isn't that bad, I wouldn't bother too much. Creating new contrast types isn't that common AFAIK.

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -0,0 +1,126 @@
+            levels::Vector{Any}
+        end
+
+        function $contrastType(v::PooledDataVector; base::Integer=1)
```
It would be more natural for `base` to be of the eltype of `v`. Though another argument, e.g. `baseind` could allow alternatively specifying it via its order.

kleinschmidt commented:
I've updated this PR with the following two big changes:

* Split off a `ContrastMatrix` type that is a contrasts + data and holds the levels, term names, and the actual matrix of contrast codes. The subtypes of `AbstractContrast` are now just containers for settings that don't depend on the data (e.g., the base index for the contrast).
* Introduce checking for non-redundant terms in model matrix construction, and "promote" non-redundant categorical variables to have full rank (#757)

nalimilan commented:
@johnmyleswhite @dmbates Any opinions about our discussion above?

nalimilan commented:
Hm, this PR completely got off the radar. Anybody willing to review?

nalimilan commented:
@kleinschmidt What's the state of this PR? Since nobody else seems to have time to review this, I think we should go ahead with the design we consider as the best one. Would you have time to resume starting on it?

kleinschmidt commented:
Oof, yeah this did totally fall of the radar (mine included). I might have some time to look at this over the weekend.

kleinschmidt commented:
I think I now see the appeal of being able to specify the levels directly in the contrast instances (vs. the `ContrastMatrix`. It would make the code a little messier (and would slightly muddy the currently clear distinction between `AbstractContrast` subtypes that are independent of data, and `ContrastMatrix` instances that depend on data), but might improve the interface from the user's point of view. This is related to the suggestion to specify the base level, rather than the base index. It would also allow for things like re-ordering contrast levels in fitting models, or limiting to only certain levels. That might be too clever for its own good though. I'll try this out and see how it feels.

Other than that, the only substantive thing was the possibility of storing contrasts in `Terms` (from `Formulas`). But those ultimately get ingested into `ModelFrame`s so that's something that's best tackled separately.

Finally, there are some cosmetic issues: how to format term names, renaming the internal `cols` function something more descriptive, and possibly other things I've missed.

I think it's worth getting this merged. Currently the generation of model matrices from formulas with categorical variables is broken and this fixes the biggest problems.

nalimilan commented on src/DataFrames.jl:
```diff
@@ -44,6 +48,7 @@ export @~,
        combine,
        complete_cases,
        complete_cases!,
+       contrast!,
```
I think the convention would be to call this `setcontrasts!`.

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -0,0 +1,233 @@
+##
+## TODO: implement contrast types in Formula/Terms
+##
+## Dave Kleinschmidt 2015-2016
```
While I really appreciate your contribution, I don't think we usually keep the contributor names in the file, as this creates unnecessary churn everytime somebody contributes to a file. git is here for that. :-)

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -0,0 +1,233 @@
+Interface to describe contrast coding schemes for categorical variables.
+
+Concrete subtypes of `AbstractContrasts` describe a particular way of converting a
+categorical column in a `DataFrame` into numeric columns in a `ModelMatrix`. Each
```
I would remove `in a DataFrame` here: I think at some point we should move this into StatsBase since it's not specific to DataFrames at all.

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -0,0 +1,233 @@
+C()                             # levels are inferred later 
+C(levels = ::Vector{Any})       # levels checked against data later
+C(base = ::Any)                 # specify base level
+C(levels = ::Vector{Any}, base = ::Any)
```
Maybe add a comment for this line too?

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -0,0 +1,233 @@
+
+## Concrete types
+
+* `TreatmentContrasts`
```
Add a short description for each type for people not familiar with these names. I also wonder whether it wouldn't be the occasion to find a more neutral name for `TreatmentContrasts`. In many situations, there's no "treatment" involved at all. Maybe `ReferenceContrasts` or `BaselineContrasts`? What terminology other software than R use?

kleinschmidt replied:
I think so. The names are extremely inconsistent across packages. I'd feel good about using the names in [this guide](http://www.ats.ucla.edu/stat/r/library/contrast_coding.htm), and providing documentation (or even `typealias`) that links those names to what they're called in R and SPSS.

nalimilan replied:
Yeah, why not? Though we should check a bit more the vocabulary used in common books on regression.

In that case, `DummyContrasts` below should be renamed to `SimpleContrasts`.

kleinschmidt replied:
I think I prefer something like `FullRankDummyContrasts` or `IndicatorContrasts` (since simple contrasts is already used to mean dummy coded contrasts where the intercept is interpreted as the grand mean).

nalimilan replied:
Looking a bit I what I could find in books and on the Web, it seems the most common terminology is "dummy/reference coding" and "effects coding" (Helmert does not change). This is what Wikipedia (judge of peace) uses: https://en.wikipedia.org/wiki/Categorical_variable#Categorical_variables_and_regression

This classification is also used by other papers: http://www.jds-online.com/files/JDS-563.pdf, http://uu.diva-portal.org/smash/get/diva2:325460/FULLTEXT01.pdf, http://statsmodels.sourceforge.net/devel/contrasts.html, http://www-psychology.concordia.ca/fac/kline/734/wendorf.pdf (I searched for "regression contrasts", a neutral expression. Also note that "Deviation coding" has much less Google results than "effects coding".)

So I would say the best terminology would be `DummyCoding`, `EffectsCoding` and `HelmertCoding`. As in the literature above, the term "contrasts" would be left for more complex cases, user-defined and specific to each analysis. As for `FullRankDummyContrasts`, I find it a bit long. Maybe just `FullDummyCoding`?

kleinschmidt replied:
Sounds good to me. I'd suggest also making type aliases for the R names (since they don't conflict with any of these names).

How about `IndicatorCoding` for the full rank version?

And I suppose it would also make sense to have a `ContrastCoding` type that takes a matrix directly. Or to expose the `ContrastsMatrix` type (it's not currently exported).

nalimilan replied:
`IndicatorCoding` isn't very explicit about how it's different from `DummyCoding`. If it's not exposed to the user, better make it long but explicit. `ContrastCoding` taking a matrix sounds good. AFAICT we need a concrete type anyway to wrap the actual matrix.

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -0,0 +1,233 @@
+* `SumContrasts`
+* `HelmertContrasts`
+
+To implement your own concrete types, implement a constructor and a method for
```
"and a method for" -> "and a `contrasts_matrix` method for"

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -0,0 +1,233 @@
+    matrix::Matrix{Float64}
+    termnames::Vector{T}
+    levels::Vector{T}
+    contrasts::AbstractContrasts
```
Replace `AbstractContrasts` with another type parameter.

kleinschmidt replied:
I worry that by doing this you'd lose information about, say, whether levels were specified on the `AbstractContrast` itself. But that might not be a problem, given that the `ContrastMatrix` type is supposed to contain all the information necessary to actually construct `ModelMatrix` columns (and it's the thing that's stored in the `ModelFrame`).

nalimilan replied:
I just meant `contrasts::T ` with `T<:AbstractContrasts`.

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -0,0 +1,233 @@
+end
+
+"""
+    ContrastsMatrix{T}(::AbstractContrasts, levels::Vector{T})
```
Missing `C::`.

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -0,0 +1,233 @@
+If levels are specified in the `AbstractContrasts`, those will be used, and likewise
+for the base level (which defaults to the first level).
+"""
+function ContrastsMatrix{T}(C::AbstractContrasts, lvls::Vector{T})
```
lvls -> levels (also for `c_lvls` inside the function).

nalimilan replied:
Also, `lvls` should be `AbstractVector`.

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -0,0 +1,233 @@
+"""
+function ContrastsMatrix{T}(C::AbstractContrasts, lvls::Vector{T})
+
+    ## if levels are defined on C, use those, validating that they line up.
```
Only one `#` (here and elsewhere).

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -0,0 +1,233 @@
+function ContrastsMatrix{T}(C::AbstractContrasts, lvls::Vector{T})
+
+    ## if levels are defined on C, use those, validating that they line up.
+    ## what does that mean?
```
To make clear you present 3 different possibilities, maybe add "either:", and bullets/numbers for each case.

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -0,0 +1,233 @@
+    ## better to filter data frame first
+    ## contrast levels missing from data: would have empty columns, generate a
+    ## rank-deficient model matrix.
+    c_lvls = convert(typeof(lvls), get(C.levels, lvls))
```
`convert(typeof(lvls), ` can be written as `oftype(lvls, `.

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -0,0 +1,233 @@
+    ## rank-deficient model matrix.
+    c_lvls = convert(typeof(lvls), get(C.levels, lvls))
+    mismatched_lvls = symdiff(c_lvls, lvls)
+    isempty(mismatched_lvls) || error("Contrasts levels not found in data or vice-versa: ", mismatched_lvls)
```
It would be useful to also print both series of levels to make it easier to find out what went wrong.

nalimilan replied:
Also, error messages should not start with an upper case (here and elsewhere).

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -0,0 +1,233 @@
+"""
+    ContrastsMatrix{T}(::AbstractContrasts, levels::Vector{T})
+
+Instantiate contrasts matrix for given data (categorical levels)
```
Maybe clearer as "Compute contrasts matrix for a given set of levels" (the actual data isn't involved at all here). Also add an ending dot.

nalimilan replied:
BTW, a comment above `ContrastsMatrix` says it's contrasts+data, I think it should say "contrasts+levels".

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -0,0 +1,233 @@
+    isempty(mismatched_lvls) || error("Contrasts levels not found in data or vice-versa: ", mismatched_lvls)
+
+    n = length(c_lvls)
+    n > 1 || error("not enough degrees of freedom to define contrasts")
```
Probably easier to understand as "at least two levels are needed to compute contrasts". It would be even better to distinguish the case where no levels are present, and the one in which only one level is present (and in that case print it).

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -0,0 +1,233 @@
+    n > 1 || error("not enough degrees of freedom to define contrasts")
+    
+    ## find index of base level. use C.base, then default (1).
+    baseind = isnull(C.base) ? 1 : findfirst(c_lvls, convert(eltype(lvls), get(C.base)))
```
AFAIK, you don't need `convert()` to check for equality using `==` (which is what `findfirst` uses).

kleinschmidt replied:
I've checked that `findfirst([:a, :b, :c], "a") = 0` which is the use case I imagined here (e.g., specify base level on contrast as a string when in fact the levels are symbols).

nalimilan replied:
OK, but this sounds like a Julia issue: https://github.com/JuliaLang/julia/issues/16271

nalimilan replied:
Could you develop why you think people might want/need to pass symbols instead of strings here?

kleinschmidt replied:
In principle, the pooled data could be of any type, and the `Contrasts` types don't have access to that when they're constructed. Of course, another possibility is to just enforce that the types of the levels/base level stored on the `Contrasts` match the levels eltype.

nalimilan replied:
Yes, just assume the types are the same, or (equivalently) leave it to the caller to use types for which `==` works.

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -0,0 +1,233 @@
+    
+    ## find index of base level. use C.base, then default (1).
+    baseind = isnull(C.base) ? 1 : findfirst(c_lvls, convert(eltype(lvls), get(C.base)))
+    baseind > 0 || error("Base level $(C.base) not found in levels")
```
Print `$levels` here too, can always be useful.

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -0,0 +1,233 @@
+    baseind > 0 || error("Base level $(C.base) not found in levels")
+    
+    not_base = [1:(baseind-1); (baseind+1):n]
+    tnames = c_lvls[not_base]
```
Is this correct for Helmert contrasts, or do we also need to allow contrasts types to provide custom names?

kleinschmidt replied:
It's correct in the sense that each contrast is interpretable as the difference between that level and the average of the preceding levels. But it might be good to allow custom names (and have a default method that behaves the current way).

nalimilan replied:
No, that's fine, we can always generalize this later.

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -0,0 +1,233 @@
+
+## Making a contrast type T only requires that there be a method for
+## contrasts_matrix(T, v::PooledDataArray). The rest is boilerplate.
+##
```
Remove empty comment line.

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -0,0 +1,233 @@
+for contrastType in [:TreatmentContrasts, :SumContrasts, :HelmertContrasts]
+    @eval begin
+        type $contrastType <: AbstractContrasts
+            base::Nullable{Any}
```
Make fields types parametric on type (including `Vector{T}` for `levels`). With this, I think you could even get rid of `nullify()`, since `convert()` should do its job automatically as long as you call `ContrastType{T}(base, levels)`.

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -0,0 +1,233 @@
+
+Needed internally when a term is non-redundant with lower-order terms (e.g., in
+~0+x vs. ~1+x, or in the interactions terms in ~1+x+x&y vs. ~1+x+y+x&y. In the
+non-redundant cases, we can (probably) expand x into length(levels(x)) columns
```
"probably" is a bit worrying here. I'd remove it, the parenthesis at the end is explicit enough. But maybe you could be more specific about what "something dumb" could be, and what could be the consequences? We don't want to return misleading results, or even let users believe this can be the case.

Also add bacticks around formulas and `length(levels(x))`.

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -0,0 +1,233 @@
+## Dummy contrasts have no base level (since all levels produce a column)
+end
+
+ContrastsMatrix{T}(C::DummyContrasts, lvls::Vector{T}) = ContrastsMatrix(eye(Float64, length(lvls)), lvls, lvls, C)
```
No need for `T` here.

nalimilan replied:
And `eye` always creates a `Float64`.

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -0,0 +1,233 @@
+
+ContrastsMatrix{T}(C::DummyContrasts, lvls::Vector{T}) = ContrastsMatrix(eye(Float64, length(lvls)), lvls, lvls, C)
+
+## Default for promoting contrasts to full rank is to convert to dummy contrasts
```
Remove these two lines?

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -0,0 +1,233 @@
+## Default for promoting contrasts to full rank is to convert to dummy contrasts
+## promote_contrast(C::AbstractContrasts) = DummyContrasts(eye(Float64, length(C.levels)), C.levels, C.levels)
+
+"Promote a contrast to full rank"
```
I'm not sure why we need `promote_contrast`. Since for now there's only one method for it, why not make it a `convert(::Type{DummyContrasts, x)` method?

nalimilan replied:
Hm, actually a `convert(::Type{ContrastsMatrix{DummyContrasts}}, x)` method (after making contrasts matrices parametric on the contrasts type).

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -0,0 +1,233 @@
+
+"Promote a contrast to full rank"
+promote_contrast(C::ContrastsMatrix) = ContrastsMatrix(DummyContrasts(), C.levels)
+
```
Two spaces should be enough... :-)

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -0,0 +1,233 @@
+"""
+    DummyContrasts
+
+One indicator (1 or 0) column for each level, __including__ the base level.
```
For this one as well as other types, I would start the sentence with something like "Contrasts type generating".

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -0,0 +1,233 @@
+
+One indicator column (1 or 0) for each non-base level.
+
+The default in R. Columns have non-zero mean and are collinear with an intercept
```
We don't really care what's the default in R. OTOH, it would be interesting to give the various names under which this contrasts type is known (reference coding, dummy coding...). Same for other types.

The first line could also explicitly mention that the base level is skipped.

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -0,0 +1,233 @@
+
+Produces mean-0 (centered) columns _only_ when all levels are equally frequent.
+But with more than two levels, the generated columns are guaranteed to be non-
+orthogonal (so beware of collinearity).
```
Not sure whether "so beware of collinearity" is really useful here.

Also, in addition to technical details, maybe start with a simpler description? E.g. "each coefficient compares the corresponding level to the average of the other levels" (for treatment, it would be "compares to the baseline", and for Helmert it would be "compares to the average of the preceding levels").

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -0,0 +1,233 @@
+    HelmertContrasts
+
+Produces contrast columns with -1 for each of n levels below contrast, n for
+contrast level, and 0 above.  Produces something like:
```
If you do this, give actual code to reproduce this example (include the julia> prompt too, cf. http://docs.julialang.org/en/release-0.4/manual/documentation/).

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -0,0 +1,233 @@
+
+This is a good choice when you have more than two levels that are equally frequent.
+Interpretation is that the nth contrast column is the difference between
+level n+1 and the average of levels 1 to n.  When balanced, it has the
```
I find it weird to be so positive about this contrast type but warn users about sum contrasts. I think we shouldn't give recommendations for users here: just highlight the properties of each type.

nalimilan commented on src/statsmodels/formula.jl:
```diff
@@ -37,6 +37,12 @@ type ModelFrame
     df::AbstractDataFrame
     terms::Terms
     msng::BitArray
+    ## mapping from df keys to contrasts.  Rather than specify types allowed,
```
The first part of the comment is good, but no need to give the rationale for the design here.

Maybe make this a `Dict{Symbol, AbstractContrasts}`?

nalimilan commented on src/statsmodels/formula.jl:
```diff
@@ -89,7 +95,7 @@ end
 
 ## Distribution of & over +
-const distributive = @compat Dict(:& => :+)
+const distributive = Dict(:& => :+)
```
Unrelated change, keep this out of this PR.

nalimilan commented on src/statsmodels/formula.jl:
```diff
@@ -234,34 +240,149 @@ function dropUnusedLevels!(da::PooledDataArray)
-function ModelFrame(trms::Terms, d::AbstractDataFrame)
+## whether or not a column of a particular type can be "promoted" to full rank
+## (only true of factors)
+can_promote(::PooledDataArray) = true
```
Maybe call this `can_promote_full` to prevent confusion with standard Julia promotion?

nalimilan replied:
Actually I now see this reflects are more fundamental property of the columns: whether they can give rise to several columns in the model matrix. I think it would be better to refer to this rather than to promotion (especially since we may need to export this to allow other types to override it). Maybe `iscategorical`?

nalimilan commented on src/statsmodels/formula.jl:
```diff
@@ -234,34 +240,149 @@ function dropUnusedLevels!(da::PooledDataArray)
+## included (either explicitly in the Terms or implicitly by promoting another
+## term like z in z&y&...).
+##
+## This function returns a boolean matrix that says whether each evaluation term
```
"evaluation term" isn't the best expression to distinguish from "term". Maybe "model matrix column"? I think this could also apply to the `termnames` field of the contrasts type (-> "colnames"?).

nalimilan replied:
Hadn't noticed that was the current terminology elsewhere in the code. Let's keep this for now then.

nalimilan commented on src/statsmodels/formula.jl:
```diff
@@ -234,34 +240,149 @@ function dropUnusedLevels!(da::PooledDataArray)
+## of each term needs to be promoted.
+function check_non_redundancy(trms::Terms, df::AbstractDataFrame)
+
+    ## This can be checked using the .factors field of the terms, which is an
```
It isn't clear to me what "this" refers to. Maybe remove the comment, since we can see from the code that this is how the function operates.

nalimilan commented on src/statsmodels/formula.jl:
```diff
@@ -234,34 +240,149 @@ function dropUnusedLevels!(da::PooledDataArray)
+                dropped[i_eterm] = 0
+                ## short circuiting check for whether the version of this term
+                ## with the current eterm dropped has been encountered already
+                ## (and hence does not need to be expanded
```
Missing closing parenthesis.

nalimilan commented on src/statsmodels/formula.jl:
```diff
@@ -234,34 +240,149 @@ function dropUnusedLevels!(da::PooledDataArray)
+        for i_eterm in 1:n_eterms
+            ## only need to check eterms that are included and can be promoted
+            ## (e.g., categorical variables that expand to multiple mm columns)
+            if round(Bool, trms.factors[i_eterm, i_term]) && can_promote(df[trms.eterms[i_eterm]])
```
Use `Bool` instead of `round`.

nalimilan commented on src/statsmodels/formula.jl:
```diff
@@ -234,34 +240,149 @@ function dropUnusedLevels!(da::PooledDataArray)
+                    end
+                end
+                ## more concisely, but with non-short-circuiting any:
+                ##is_redundant = any([dropped == enc for enc in encountered_columns])
```
Isn't this simply `findfirst(encountered_columns, dropped) != 0`?

nalimilan commented on src/statsmodels/formula.jl:
```diff
@@ -234,34 +240,149 @@ function dropUnusedLevels!(da::PooledDataArray)
+                end
+
+            end
+            ## once we've checked all the eterms in this term, add it to the list
```
Looks like this comment needs to be move one line below.

nalimilan commented on src/statsmodels/formula.jl:
```diff
@@ -234,34 +240,149 @@ function dropUnusedLevels!(da::PooledDataArray)
+    end
+
+    return to_promote
+    
```
Remove empty line.

nalimilan commented on src/statsmodels/formula.jl:
```diff
@@ -234,34 +240,149 @@ function dropUnusedLevels!(da::PooledDataArray)
+evaluateContrasts{C <: AbstractContrasts}(c::Type{C}, col::AbstractDataVector) = ContrastsMatrix(c(), col)
+evaluateContrasts(c::ContrastsMatrix, col::AbstractDataVector) = c
+
+needsContrasts(::PooledDataArray) = true
```
Looks like this could be merged with the `iscategorical` function I proposed above.

nalimilan commented on src/statsmodels/formula.jl:
```diff
@@ -234,34 +240,149 @@ function dropUnusedLevels!(da::PooledDataArray)
+                                                  col)
+    end
+
+    ## Check whether or not
```
Not very useful comment.

nalimilan commented on src/statsmodels/formula.jl:
```diff
@@ -234,34 +240,149 @@ function dropUnusedLevels!(da::PooledDataArray)
+## Goal here is to allow specification of _either_ a "naked" contrast type,
+## or an instantiated contrast object itself.  This might be achieved in a more
+## julian way by overloading call for c::AbstractContrasts to just return c.
+evaluateContrasts(c::AbstractContrasts, col::AbstractDataVector) = ContrastsMatrix(c, col)
```
Can't you turn these into methods for `ContrastMatrix`?

nalimilan commented on src/statsmodels/formula.jl:
```diff
@@ -234,34 +240,149 @@ function dropUnusedLevels!(da::PooledDataArray)
+
+## modify contrasts in place
+function contrast!(mf::ModelFrame, new_contrasts::Dict)
+    new_contrasts = [ col => evaluateContrasts(contr, mf.df[col])
```
Could raise an error here if the dict contains references to columns that are not present in the data frame. Then you wouldn't need to call `filter`.

nalimilan commented on src/statsmodels/formula.jl:
```diff
@@ -234,34 +240,149 @@ function dropUnusedLevels!(da::PooledDataArray)
+ModelFrame(ex::Expr, d::AbstractDataFrame; kwargs...) = ModelFrame(Formula(ex), d; kwargs...)
+
+## modify contrasts in place
+function contrast!(mf::ModelFrame, new_contrasts::Dict)
```
I guess this should be documented and exported, right?

nalimilan commented on src/statsmodels/formula.jl:
```diff
@@ -234,34 +240,149 @@ function dropUnusedLevels!(da::PooledDataArray)
+    mf.contrasts = merge(mf.contrasts, new_contrasts)
+    return mf
+end
+contrast!(mf::ModelFrame; kwargs...) = contrast!(mf, Dict(kwargs))
```
Maybe a more natural API would be that the method above *replaces* the contrasts, leaving unspecified columns use the default (treatment) contrasts, while this method would only modify the specified ones? Else, you wouldn't have an easy way of resetting contrasts. Not sure.

nalimilan commented on src/statsmodels/formula.jl:
```diff
@@ -234,34 +240,149 @@ function dropUnusedLevels!(da::PooledDataArray)
+    ## reindex with v.refs will give the corresponding row number of the
+    ## contrast matrix
+    reindex = [findfirst(contrast.levels, l) for l in levels(v)]
+    return contrast.matrix[reindex[v.refs], :]
```
So this may change the order of appearance of columns and coefficients? Doesn't sound like a good idea, especially for Helmert contrasts.

nalimilan commented on src/statsmodels/formula.jl:
```diff
@@ -291,57 +412,105 @@ function nc(trm::Vector)
+    eterm_cols = Dict{Tuple{Symbol,Bool}, Array{Float64}}()
+    ## Accumulator for each term's vector of eval. term columns.
+    mm_cols = Vector{Array{Float64}}[]
+    mf.terms.intercept && push!(mm_cols, Matrix{Float64}[ones(Float64, size(mf.df,1), 1)])
```
I don't understand why you need `Matrix{Float64}[` here.

nalimilan commented:
Sorry for the delay. This looks really cool, though I've made a lot of comments. I still need to continue reading it to see the bigger picture.

nalimilan commented on src/statsmodels/formula.jl:
```diff
@@ -291,57 +412,105 @@ function nc(trm::Vector)
+    ## "promoted" full-rank versions of categorical columns for non-redundant
+    ## eval. terms:
+    for (i_term, term) in enumerate(mf.terms.terms)
+        ## Skip non-fixed-effect terms
```
This kind of comment isn't very useful when the code is explicit enough.

nalimilan commented on src/statsmodels/formula.jl:
```diff
@@ -291,57 +412,105 @@ function nc(trm::Vector)
+        non_redundant = mf.non_redundant_terms[factors[:, i_term], i_term]
+        ## Get cols for each eval term (either previously generated, or generating
+        ## and storing as necessary)
+        for et_and_nr in zip(eterms, non_redundant)
```
`et_and_nr` is an unusual name for a variable in the code base. I would find it clearer to do `for (et, nr)`, and pair them where necessary.

nalimilan commented on src/statsmodels/formula.jl:
```diff
@@ -291,57 +412,105 @@ function nc(trm::Vector)
+        ## and storing as necessary)
+        for et_and_nr in zip(eterms, non_redundant)
+            haskey(eterm_cols, et_and_nr) || 
+                setindex!(eterm_cols,
```
Clearer with an `if` block, and as `eterm_cols[i] = et_and_nr`, with `i = modelmat_cols(et_and_nr[1], mf, non_redundant=et_and_nr[2])`.

nalimilan commented on src/statsmodels/formula.jl:
```diff
@@ -291,57 +412,105 @@ function nc(trm::Vector)
+
+    ## TODO: could this be made more efficient by
+    ## first computing mm_col_term_nums, initializing mm, and directly indexing?
+    mm = hcat([expandcols(tc) for tc in mm_cols]...)
```
Splatting should never be done on vectors whose size is data-dependent, as this will require compiling specific methods for each length. Better use a comprehension here.

kleinschmidt replied:
Could you elaborate? Also, this is a bit orthogonal to the contrast coding, so maybe better left for another PR.

nalimilan replied:
Well, it's new code AFAICT, so it's relevant to this PR. What I meant is that splatting will force compiling `hcat` for each type/number combination of its argument. Since these vary a lot depending on the data, this will trigger the compilation of many functions. See https://github.com/JuliaLang/julia/issues/13722

So indeed the TODO is right that it would be better to create the matrix first, and then fill it with a loop. Not sure whether there's a simpler solution.

kleinschmidt replied:
Okay, I see the problem. That part's not new though, it replaces this code using more verbose variable names:

    hcat([expandcols(t) for t in aa]...)

nalimilan commented on src/statsmodels/formula.jl:
```diff
@@ -291,57 +412,105 @@ function nc(trm::Vector)
+    mm_col_term_nums = vcat([fill(i_term, nc(tc)) for (i_term,tc) in enumerate(mm_cols)]...)
+
+    ModelMatrix{Float64}(mm, mm_col_term_nums)
+
```
Remove empty line.

nalimilan commented on src/statsmodels/formula.jl:
```diff
@@ -291,57 +412,105 @@ function nc(trm::Vector)
+    ## TODO: could this be made more efficient by
+    ## first computing mm_col_term_nums, initializing mm, and directly indexing?
+    mm = hcat([expandcols(tc) for tc in mm_cols]...)
+    mm_col_term_nums = vcat([fill(i_term, nc(tc)) for (i_term,tc) in enumerate(mm_cols)]...)
```
Same here (use `vec` on a comprehension)`.

nalimilan replied:
Same here.

nalimilan commented on src/statsmodels/formula.jl:
```diff
@@ -291,57 +412,105 @@ function nc(trm::Vector)
-function termnames(term::Symbol, col::PooledDataArray)
-    levs = levels(col)
-    [string(term, " - ", levs[i]) for i in 2:length(levs)]
+function termnames(term::Symbol, mf::ModelFrame; non_redundant::Bool = false)
```
The double negation with `non_redundant` isn't great. Maybe better say "full" or "fullrank" here and everywhere else?

nalimilan commented on src/statsmodels/formula.jl:
```diff
@@ -291,57 +412,105 @@ function nc(trm::Vector)
+function termnames(term::Symbol, mf::ModelFrame; non_redundant::Bool = false)
+    if haskey(mf.contrasts, term)
+        termnames(term, mf.df[term], 
+             non_redundant ? promote_contrast(mf.contrasts[term]) : mf.contrasts[term])
```
Align with other arguments.

nalimilan commented on src/statsmodels/formula.jl:
```diff
@@ -291,57 +412,105 @@ function nc(trm::Vector)
+        for et_and_nr in zip(eterms, non_redundant)
+            haskey(eterm_names, et_and_nr) ||
+            setindex!(eterm_names,
+                      termnames(et_and_nr[1], mf, non_redundant=et_and_nr[2]),
```
Calling `termnames` repeatedly doesn't sound like the best strategy, since most of the work is common to all evaluated terms belonging to the same term/variable. Couldn't you find a better abstraction model here? Also, since you note the logic of this function is similar to `ModelMatrix`, maybe some parts could be factored out?

kleinschmidt replied:
This only calls `termnames` once for each eval term (column in the data frame), or twice if it occurs in both redundant and non-redundant forms, regardless of how many terms it appears in.

That being said, it might make sense to calculate the coef names once and store them on the model matrix itself if we switch to something like a `NamedArray`. Then some of this could be refactored into the same loop. But the issue then would be that `coefnames` lives on the model matrix, whereas now it lives on the model frame, so the interface would have to change. That would (as far as I can tell) be pretty minimal, since the only place we use it is in `coeftable` for `DataFrameModels`, which hold the `ModelMatrix` and the `ModelFrame`.

nalimilan replied:
OK. I think we should store them both in the model frame and as column names in a `NamedArray` model matrix, but this can be left for another PR.

Why are you using `setindex!` instead of indexing?

kleinschmidt replied:
I don't remember at this point. I switched them both to use indexing in a commit I pushed last night...

nalimilan commented on src/statsmodels/formula.jl:
```diff
@@ -291,57 +412,105 @@ function nc(trm::Vector)
+        eterms = mf.terms.eterms[factors[:, i_term]]
+        non_redundant = mf.non_redundant_terms[factors[:, i_term], i_term]
+
+        for et_and_nr in zip(eterms, non_redundant)
```
Here too, `for (et, nr)` would be more natural.

nalimilan commented on src/statsmodels/formula.jl:
```diff
@@ -291,57 +412,105 @@ function nc(trm::Vector)
-    return vnames
+
+    mapreduce(expandtermnames, vcat, term_names)
+    
```
Remove empty line.

nalimilan commented on src/statsmodels/formula.jl:
```diff
@@ -291,57 +412,105 @@ function nc(trm::Vector)
+
+    factors = round(Bool, mf.terms.factors)
+
+    for (i_term, term) in enumerate(mf.terms.terms)
```
Here and elsewhere, I think `i` would be better than `i_term` since there's no ambiguity.

kleinschmidt replied:
There's not ambiguity in principle, but I (at least) find it helps in reading the code to know when we're indexing terms vs. evalterms or something else. But perhaps that's just lingering PTST from all the single-letter index/variable names that used to be in these functions...

nalimilan commented on test/contrasts.jl:
```diff
@@ -0,0 +1,92 @@
+                            1 -1 -1
+                            1 -1 -1
+                            1  1  0]
+@test coefnames(mf) == ["(Intercept)"; "x: b"; "x: c"]
```
For sum contrasts, I think it is more common to omit the *last* level by default. Do you think it would be possible without too much work?

kleinschmidt replied:
That is in fact what R does. But I've always found it a little odd that `contr.treatment` and `contr.helmert` go the other way. So while it wouldn't be technically that difficult to change, I'd prefer to keep the default for the base level as consistent as possible across types.

nalimilan replied:
I see. Let's keep it that way then. Anyway I'd like to also compute the redundant coefficient which wasn't included in the matrix, which is clearly useful here.

nalimilan commented on test/contrasts.jl:
```diff
@@ -0,0 +1,92 @@
+                            1  1 -1]
+@test coefnames(mf) == ["(Intercept)"; "x: b"; "x: c"]
+
+## Types for contrast levels are coerced to data levels when constructing
```
"coerced" isn't a Julia term. Do you mean "converted"? If so, then this test doesn't actually check it.

nalimilan commented on test/contrasts.jl:
```diff
@@ -0,0 +1,92 @@
+
+## Missing data is handled gracefully, dropping columns when a level is lost
+d[3, :x] = NA
+mf_missing = ModelFrame(Formula(Nothing(), :x), d, contrasts = Dict(:x => SumContrasts))
```
Why not also test the default case? BTW, also check that Helmert contrasts behave correctly with missing data.

nalimilan commented on test/contrasts.jl:
```diff
@@ -0,0 +1,92 @@
+## Applying contrasts that only have a subset of data levels:
+@test_throws ErrorException contrast!(mf, x = SumContrasts(levels = [:a, :b]))
+## Applying contrasts that expect levels not found in data:
+@test_throws ErrorException contrast!(mf, x = SumContrasts(levels = [:a, :b, :c, :d]))
```
Could also test trying to set the base level to a non-existent level.

nalimilan commented on test/formula.jl:
```diff
@@ -365,4 +368,38 @@ module TestFormula
+
+
+## No intercept
+mf = ModelFrame(n ~ 0 + x, d[1:2,:], contrasts=cs)
```
Why do you select only a subset of the data here?

kleinschmidt replied:
This test is for promoting one variable to full rank when the intercept is missing. It could be re-written using a separate dataframe for each test if that would be clearer...

nalimilan replied:
I still don't get it. What's the difference here, since the levels of `x` are the same? Also, here and below, you only use the first 4 rows, so why create more of them?

kleinschmidt replied:
I just didn't want to type out the additional rows of the model matrix.

I think I intended to test three-way and multiple two-way interactions (with different patterns of redundancy). I'm not sure that's necessary at this point, but there's some value in it at least from the perspective of being explicit about the expected behavior.

nalimilan replied:
I think it would be clearer to use all rows, even it this indeed requires more typing. Also remove rows above 4, and `[1:4, :]`.

Testing complex cases would indeed be a good idea while you're at it. I'm afraid we won't return soon to this part of the code, in particular to check corner cases...

nalimilan commented on test/formula.jl:
```diff
@@ -314,27 +314,30 @@ module TestFormula
-             y in 5:6,
-             z in 8:9])
+    ## 
+    ## FAILS: behavior is wrong when no lower-order terms (1+x1+x2+x1&x2...)
```
Do you have a plan to fix this?

kleinschmidt replied:
I think I meant that the behavior the test expects is wrong (two of them should be promoted to full rank, I think. I'll have to double check though). Other tests for missing lower-order terms check this behavior though. Probably best to just remove this one.

nalimilan replied:
Indeed, if it's incorrect, remove it. Though double check with e.g. R that this is the standard behavior.

nalimilan commented:
While we're at it, I think we should remove the `ModelMatrix` type and simply return a `NamedArray`. Returning an array without any way of identifying columns is a very bad API IMHO.

kleinschmidt commented:
Thanks for the super thorough review, @nalimilan. I've addressed a lot of your suggestions in the commits I just pushed. There's still a bit more to do in settling on and documenting the names for things, although I might not get around to that till the weekend.

nalimilan commented:
Sorry for the delay. I'll try to find the time review this again, but I think we're coming close.

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -116,12 +116,22 @@ function ContrastsMatrix{C <: AbstractContrasts}(contrasts::C, levels::Vector)
+    isempty(setdiff(levels(col), c.levels)) ?
     c :
-    error("mismatch between levels in ContrastsMatrix and data:\nData levels: $(levels(col))\nContrast levels $(c.levels)")
+    error("there are levels in data that are not in ContrastsMatrix: $(setdiff(levels(col), c.levels))\n  Data levels: $(levels(col))\n  Contrast levels $(c.levels)")
```
I'd rather write this as "data contains levels not in ContrastsMatrix:" or "levels not in ContrastsMatrix found in data:".

nalimilan replied:
I also wonder whether it's a good idea to add indent to the two last lines. You don't follow the same convention in `ContrastsMatrix`.

nalimilan commented on test/formula.jl:
```diff
@@ -379,27 +379,98 @@ d[:n] = 1.:8
+
+
+
 ## FAILS: When both terms are non-redundant and intercept is PRESENT
```
Just remove these commented tests if they are incorrect. You seem to have considered the question much more carefully than what had been done before.

Also, two empty lines are enough in general.

nalimilan replied:
Ah, I just realize that you added this test (I thought it was an old one). Let's keep it then, but if you don't think you can fix it soon open an issue so that we don't forget about it.

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -1,40 +1,38 @@
+# ModelMatrix will check this dict when evaluating terms, falling back to a
+# default for any categorical data without a specified contrast.
+#
+# TODO: implement contrast types in Formula/Terms
```
Not sure leaving a TODO here is a good idea. You might well forget it when implementing it since it's not in the same file...

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -50,15 +48,15 @@ level.
-constructing the actual contrasts matrix that maps from levels to `ModelMatrix`
-column values:
+To implement your own concrete types, implement a constructor and a
+`contrast_matrix` method for constructing the actual contrasts matrix
```
"contrasts_matrix"

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -71,49 +69,50 @@ contrasts_matrix(C::MyContrasts, baseind, n) = ...
+
+    n = length(c_levels)
+    n == 0 && error("empty set of levels found (need at least two to compute contrasts).")
+    n == 1 && error("only one level found: $(c_levels[1]). need at least two to compute contrasts.")
```
Upper case "N". There's an inconsistency in the punctuation w.r.t. the message above.

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -71,49 +69,50 @@ contrasts_matrix(C::MyContrasts, baseind, n) = ...
     not_base = [1:(baseind-1); (baseind+1):n]
     tnames = c_lvls[not_base]
+    # find index of base level. use C.base, then default (1).
+    baseind = isnull(C.base) ? 1 : findfirst(c_levels, convert(eltype(levels), get(C.base)))
```
I though we decided to get rid of the `convert(eltype(levels), ...)` stuff?

Also, a single empty line is enough inside a function.

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -0,0 +1,245 @@
+for contrastType in [:TreatmentContrasts, :SumContrasts, :HelmertContrasts]
+    @eval begin
+        type $contrastType <: AbstractContrasts
+            base::Nullable{Any}
```
I think you should restrict the types of the fields. `levels` should be `Nullable{Vector{T}}` with `T` a type parameter, and `base` should be a `Nullable{Int}` (or `Nullable{T}`). Apart from removing type instability, that should help making it clearer which types are expected in the rest of the code.

kleinschmidt replied:
I've thought a bit about this and I'm not so sure. For one, if we only allow instantiate contrasts types to be passed to `ModelFrame` et al., that would force users to actually specify the type of the column when they specify the contrasts. I'm also not sure that type instability would actually have any substantial performance costs here, since all of the heavy lifting is done by the `ContrastsMatrix` type which _is_ parametrized on the element type of the column it is created from (or rather the eltype of the levels).

nalimilan replied:
Fair enough.

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -85,33 +85,35 @@ Compute contrasts matrix for a given set of categorical data levels.
     # 3. contrast levels missing from data: would have empty columns, generate a
     #    rank-deficient model matrix.
-    c_levels = oftype(levels, get(C.levels, levels))
+    c_levels = oftype(levels, get(contrasts.levels, levels))
```
Remove the conversion here too.

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -85,33 +85,35 @@ Compute contrasts matrix for a given set of categorical data levels.
+    # find index of base level. use contrasts.base, then default (1).
+    baseind = isnull(contrasts.base) ?
+              1 :
+              findfirst(c_levels, convert(eltype(levels), get(contrasts.base)))
```
And here.

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -160,14 +160,8 @@ end
 
 ContrastsMatrix{T}(C::DummyContrasts, lvls::Vector{T}) = ContrastsMatrix(eye(Float64, length(lvls)), lvls, lvls, C)
```
Cut the lines at 92 chars (here and elsewhere).

nalimilan commented on src/statsmodels/formula.jl:
```diff
@@ -419,12 +419,11 @@ function ModelMatrix(mf::ModelFrame)
-                          et_and_nr)
-            push!(term_cols, eterm_cols[et_and_nr])
+        for (et, nr) in zip(eterms, non_redundant)
+            if ! haskey(eterm_cols, (et, nr))
```
No space after `!`.

nalimilan commented on src/statsmodels/contrasts.jl:
```diff
@@ -117,6 +117,11 @@ function ContrastsMatrix{C <: AbstractContrasts}(contrasts::C, levels::Vector)
 ContrastsMatrix(C::AbstractContrasts, v::PooledDataArray) = ContrastsMatrix(C, levels(v))
+ContrastsMatrix{C <: AbstractContrasts}(c::Type{C}, col::PooledDataArray) = ContrastsMatrix(c(), col)
+ContrastsMatrix(c::ContrastsMatrix, col::PooledDataArray) =
+    isempty(symdiff(c.levels, levels(col))) ?
```
The ternary operator isn't the best solution when you need to add line breaks. You can write this as:
```julia
isempty(symdiff(c.levels, levels(col))) || error(...)
return c
```

Also, in other places you explicitly list the mismatched contrasts.

nalimilan commented on src/statsmodels/statsmodel.jl:
```diff
@@ -48,8 +48,8 @@ for (modeltype, dfmodeltype) in ((:StatisticalModel, DataFrameStatisticalModel),
         function StatsBase.fit{T<:$modeltype}(::Type{T}, f::Formula, df::AbstractDataFrame,
-                                              args...; kwargs...)
-            mf = ModelFrame(f, df)
+                                              args...; contrasts::Dict = Dict(), kwargs...)
```
Maybe `Dict{Symbol, C}` with `C<:AbstractContrasts`?

kleinschmidt replied:
This would preclude using un-instantiated contrasts (e.g., `HelmertCoding` rather than `HelmertCoding()`) which is currently supported. If you think that's too clever we could restrict it like you suggest, but it's an idiom I borrowed from Gadfly.jl.

nalimilan replied:
In general I think the recommended pattern in Julia is to pass instances rather than types, which is e.g. how specifying a distribution works. So if it's just to avoid typing `()`, I'd rather not support this variant to keep things simple. We can always add it later if it really turns out to be useful.

kleinschmidt replied:
I see the point there, and agree there's really no benefit to allowing `HelmertCoding` etc. I've changed the contrasts matrix method for `Type{C<:AbstractContrasts}` to give an informative error message.

Let's restrict this dict to `AbstractContrasts` for now and then see if other use cases pop up. I can imagine someone might want to construct a `ContrastsMatrix` directly (perhaps re-using from a previously fitted model's `ModelFrame`) but my intuition is to keep things simple here and introduce more complexity as needed.

kleinschmidt replied:
Actually, now I'm confused:

```julia
julia> d = Dict(:x1p => EffectsCoding(),  :x2p => DummyCoding())
Dict{Symbol,Any} with 2 entries:
  :x2p => DataFrames.DummyCoding(Nullable{Any}(),Nullable{Array{T,1}}())
  :x1p => DataFrames.EffectsCoding(Nullable{Any}(),Nullable{Array{T,1}}())

julia> d::Dict{Symbol, AbstractContrasts}
ERROR: TypeError: typeassert: expected Dict{Symbol,DataFrames.AbstractContrasts}, got Dict{Symbol,Any}

```

So restricting the dict type with `contrasts::Dict{Symbol, AbstractContrasts}` throws an error in this (perfectly reasonable) case, even though you can convert fine:

```julia
julia> convert(Dict{Symbol, AbstractContrasts}, d)
Dict{Symbol,DataFrames.AbstractContrasts} with 2 entries:
  :x2p => DataFrames.DummyCoding(Nullable{Any}(),Nullable{Array{T,1}}())
  :x1p => DataFrames.EffectsCoding(Nullable{Any}(),Nullable{Array{T,1}}())
```

nalimilan replied:
That's an issue with type inference at the REPL. If people really need to do this kind of thing, then probably better keep the signature looser then. We could still check manually that all values are of `AbstractContract` subtypes.

kleinschmidt replied:
I don't think it's just a REPL problem; I got this error initially when I wrote a test for the two-contrast case there. 

Also, the check happens when the `ModelFrame` is constructed, because it converts all the contrasts to `ContrastsMatrix` instances, which only has methods for subtypes of `AbstractContrasts` and `ContrastsMatrix` objects. I suppose the downside of only checking then is that you don't get information in the error message about the variable that contrast belongs to, but that's something we can address after merging if need be.

nalimilan commented:
Do you think you'll be able to fix the remaining points? Let's not leave this PR rot again for months. ;-)

kleinschmidt commented:
Sorry to let this slide a bit. I'm writing my dissertation at the moment. I'll try to squeeze this in but I'm pretty swamped for the next few weeks. 



> On Jun 3, 2016, at 10:56 AM, Milan Bouchet-Valat <notifications@github.com> wrote:
> 
> Do you think you'll be able to fix the remaining points? Let's not leave this PR rot again for months. ;-)
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

nalimilan commented:
No worries -- I know what this means...

nalimilan commented:
Do you think the PR could be merged in its current state, or maybe with only minimal fixes? Then we could address stylistic comments and performance improvements later. It would be a shame to let this blocked by details.

kleinschmidt commented:
I think we should hold off on merging this in light of #1017. Also, as discussed on the juliastats list, it's not clear whether DataFrames.jl is the best home for this functionality. On the one hand, it seems more natural to define how to turn categorical variables into model matrix columns in the packages that define the categorical types themselves. On the other hand, I think its good to define the different contrast types themselves in a way that's agnostic to the underlying categorical data representation. Moreover, there are issues in constructing a whole modelmatrix with categorical data that go beyond converting individual categorical variables into columns (e.g., checking for redundancy with lower-order terms). So there's at least some aspects of this PR that _should_, I think, be included here, and yet others that might be factored out into a contrasts-specific package that others can use to convert categorical data into matrices without needing to put them in a Data/ModelFrame.

Thoughts? @nalimilan @dmbates @ararslan

ararslan commented:
I was just thinking about this the other day. The modeling stuff (including formulas) seem out of place here; there's nothing inherently statistical about a data frame. IMO a package for data frames should provide the data frame type, which other packages can then extend for particular purposes. See also #1003.

kleinschmidt commented:
Right. I'm starting to think it makes sense to have something like a
DataFrameModels package that requires little self-contained packages for
Formulas and Contrasts, in addition to DataFrames, and defines the
DataFrameRegressionModel etc. types.

On Thu, Jul 21, 2016 at 3:19 PM, Alex Arslan <notifications@github.com>
wrote:

> I was just thinking about this the other day. The modeling stuff
> (including formulas) seem out of place here; there's nothing inherently
> statistical about a data frame. IMO a package for data frames should
> provide the data frame type, which other packages can then extend for
> particular purposes. See also #1003
> <https://github.com/JuliaStats/DataFrames.jl/issues/1003>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaStats/DataFrames.jl/pull/870#issuecomment-234355686>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAIS8PRdRLD6PNv_EefvbgElfVfre2wCks5qX8YqgaJpZM4GApdq>
> .
>



-- 
Dave Kleinschmidt
Brain and Cognitive Sciences, University of Rochester
http://www.bcs.rochester.edu/people/dkleinschmidt

dmbates commented:
@kleinschmidt Agree re small packages used by `DataFramesModels`, although I might emphasize the role of `ModelMatrix` rather than the `DataFrames` aspect of things.  What the whole formula/data representation is about is transitioning from a formula specification plus some kind of associative structure mapping names to vectors to a model matrix of some form.  In R formulas ended up being used for other purposes but I think in modeling the vast majority of uses are for creating a model matrix from a formula.  You don't need a DataFrame specifically for that.  You could construct from a database table or a feather file directly.

kleinschmidt commented:
I've opened an issue #1018 to discuss this; let's move this discussion there?

nalimilan commented:
I think we should finish and merge this and #1017, and then move the code (preserving git history) to another package as discussed at #1018.

kleinschmidt commented:
That seems reasonable to me, @nalimilan. I'll wait until #1017 is merged, update this, fix the outstanding issues, and then we can merge this and think about refactoring.

kleinschmidt commented:
I'm in the process of reconciling this with #1017 before I finish the last few things we talked about here. But I wanted to make a note that I'm more and more convinced that we should be storing contrast information on the Terms object itself. It would substantially clean up the model matrix construction; the strategy from #1017 could be used basically without modification if we annotate the terms/evaluation terms with the contrast information (including whether particular evaluation terms need to be promoted to full-rank dummy contrasts). This would require modifying the Terms object itself when constructing a ModelFrame, and would also (potentially) make the ModelFrame itself kind of redundant, so it's probably something to hold off on until we refactor the modeling stuff out of DataFrames.

kleinschmidt commented:
I think this is about ready to merge. The last thing to settle is whether to be more specific about types in a few places (I've replied to line comments about that above). @nalimilan, want to take one more look?

kleinschmidt commented:
(Test failures on appveyor are unrelated, I think; failing in RDA.jl)

nalimilan commented:
I won't have the time to review this again, but since you addressed my comments, I think this can be merged. Do you think you could squash this into a few meaningful separate commits, or would it be too much work?

kleinschmidt commented:
Okay, thanks for all your careful attention to this! I'll merge tomorrow unless anyone else objects. @johnmyleswhite, @ararslan, @dmbates, speak now or forever hold your peace.

ararslan commented:
My peace will forever be held. Looks good to me! Great work here!

dmbates commented:
LGTM

kleinschmidt commented:
@nalimilan it's going to be a pain in the ass to squash these commits manually. Is it okay to use the autosquash "Squash and merge" option?

nalimilan commented:
Yes, that's fine since all changes are logically related.

johnmyleswhite commented:
I just wanted to say thanks for doing such an impressive job with this, @kleinschmidt. This was one of the most impressive PR's I've ever seen in a JuliaStats repo.

kleinschmidt commented:
Thanks, it was my pleasure! Lots more to work on here :)
